pkg/providers/launchtemplate/launchtemplate.go

```diff
@@ -148,16 +151,25 @@ func (p *Provider) createAMIOptions(ctx context.Context, nodeTemplate *v1alpha1.
 	if len(securityGroupsIDs) == 0 {
 		return nil, fmt.Errorf("no security groups exist given constraints")
 	}
+	if err != nil {
+		return nil, err
+	}
+	anyPublicIPv4Associations, err := p.subnetProvider.CheckAnyPublicIPv4Associations(ctx, nodeTemplate)
+	if err != nil {
+		return nil, err
+	}
+
 	return &amifamily.Options{
```

Comment:
How do you feel about being really explicit about how this is being set in code combined with a comment on why we are doing this explicit set logic? i.e. you do the following in code

```go
options := amifamily.Options{...} // leave the AssociatePublicIPAddress nil
if ok, err := p.subnetProvider.CheckAnyPublicIPAssociations(ctx, nodeTemplate); err != nil {
  return nil, err
} else if !ok {
  // provide some comment here on why we are doing this
  options.AssociatePublicIpAddress = aws.Bool(false)
}
return options
```

Reply:
> The snippet above changes the logic, and in the wrong place.

How do you mean that it changes the logic in the wrong place?

1. I would expect that the concept of `checkAnyPublicIPAddresses` would return `true` if any IPs have a public mapping, I think that makes sense as we have it.
2. I would expect that the value `AssociatePublicIpAddress` in the `amifamily.Options` would appropriately map to the same value in the `networkInterface` (since it has the same name). In this case, we are _not_ doing that and I think that it's likely to cause some confusion. To me it makes more sense to handle setting that value to whatever it should be, outside the context of the options, so that you properly abstract the handling of passing that value down to `generateNetworkInterfaces`. i.e. `generateNetworkInterfaces` doesn't leak logic into this feature. If things change later and we do something different, like surface `v1alpha1.NetworkInterfaces` through the API, `generateNetworkInterfaces` doesn't have to change since it still handles passing the `AssociatePublicIpAddress` value through in an abstracted, non-leaky way.